src/general/LimitOrderHook.sol

```diff
@@ -0,0 +1,797 @@
+        returns (int24 tickLower, int24 lower, int24 upper)
+    {
+        tickLower = getTickLower(getTick(poolId), tickSpacing);
+        int24 tickLowerLast = getTickLowerLast(poolId);
+
+        if (tickLower < tickLowerLast) {
+            lower = tickLower + tickSpacing;
+            upper = tickLowerLast;
+        } else {
+            lower = tickLowerLast;
+            upper = tickLower - tickSpacing;
+        }
+    }
+
+    /**
+     * @dev Get the last tick lower.
+     *
+     * @param poolId The pool id.
+     * @return tickLowerLast The last tick lower.
+     */
+    function getTickLowerLast(PoolId poolId) public view returns (int24) {
+        return tickLowerLasts[poolId];
+    }
+
+    /**
+     * @dev Set the last tick lower.
+     *
+     * @param poolId The pool id.
+     * @param tickLower The tick lower.
+     */
+    function setTickLowerLast(PoolId poolId, int24 tickLower) private {
+        tickLowerLasts[poolId] = tickLower;
+    }
+
+    /**
+     * @dev Get the epoch for a given pool and tick.
+     *
+     * @param key The pool key.
+     * @param tickLower The lower tick.
+     * @param zeroForOne Whether the limit order is for buy `currency0` or `currency1`.
+     * @return epoch The epoch.
+     */
+    function getEpoch(PoolKey memory key, int24 tickLower, bool zeroForOne) public view returns (Epoch) {
+        return epochs[keccak256(abi.encode(key, tickLower, zeroForOne))];
+    }
+
+    /**
+     * @dev Set the epoch for a given pool and tick.
+     *
+     * @param key The pool key.
+     * @param tickLower The lower tick.
+     * @param zeroForOne Whether the limit order is for buy `currency0` or `currency1`.
+     * @param epoch The epoch.
+     */
+    function setEpoch(PoolKey memory key, int24 tickLower, bool zeroForOne, Epoch epoch) private {
+        epochs[keccak256(abi.encode(key, tickLower, zeroForOne))] = epoch;
+    }
+
+    /**
+     * @dev Get the tick lower.
+     *
+     * @param tick The tick.
+     * @param tickSpacing The tick spacing.
+     * @return tickLower The lower tick.
+     */
+    function getTickLower(int24 tick, int24 tickSpacing) private pure returns (int24) {
+        int24 compressed = tick / tickSpacing;
+        if (tick < 0 && tick % tickSpacing != 0) compressed--; // round towards negative infinity
+        return compressed * tickSpacing;
+    }
+
+    /**
+     * @dev Get the epoch liquidity for a given epoch and owner.
+     *
+     * @param epoch The epoch.
+     * @param owner The owner.
+     * @return liquidity The liquidity.
+     */
+    function getEpochLiquidity(Epoch epoch, address owner) external view returns (uint256) {
+        return epochInfos[epoch].liquidity[owner];
+    }
+
+    /**
+     * @dev Get the tick for a given pool.
+     *
+     * @param poolId The pool id.
+     * @return tick The tick.
+     */
+    function getTick(PoolId poolId) private view returns (int24 tick) {
+        (, tick,,) = poolManager.getSlot0(poolId);
+    }
```

Comment:
this is unsafe in very obscure cases -- cetora found it in the v4 formal verification, and we refer to it as M02 behavior

for zeroForOne swaps, which moves `slot0.sqrtPriceX96` to exactly a `tickLower` of an existing position, `slot0.tick` is decremented by one

https://github.com/Uniswap/v4-core/pull/930/files#diff-d198b8552afdb46e50c8452630825f118160ea6f2da94677d8e6d48d7ceeac95R1231-R1243

---

The safest way to obtain a tick is:

```solidity
(uint160 sqrtPriceX96,,,) = poolManager.getSlot0(poolId);
tick = TickMath.getTickAtSqrtPrice(sqrtPriceX96);
```